## IIP314W Optimización aplicada a negocios

Metodo Simplex - Matricial

### Problema de Optimización General

$$ Min \quad c \cdot x $$

$$s.a, \quad A x \le b$$

$$ x \ge 0$$

Ya vimos la forma de resolución de este problema mediante el algoritmo simplex tableau, ahora realizaremos lo mismo mediante el uso de operaciones matriciales. Al igual que en el caso anterior, partimos ingresando las variables de holgura, entonces al inicio tenemos lo siguiente:

| V.Básica | Ecuación | $\vec{x}$ | $\vec{x_B}$ | LD |
|:-:|:-:|:-:|:-:|:-:|
| z | F.Obj. | $c$ | 0 | 0 |
| $\vec{x_B}$ | Rest. | $A$ | $I$ | $b$ |

Donde $x_B$ son las variables de holgura creadas en el inicio. Veamos si es óptimo.

#### Prueba de optimalidad

$$ c_B \cdot B^{-1} A - c $$

si algún coeficiente es negativo (para máximo, solo cambia el signo para minimo), la solución no es óptima. En nuestro caso, $c_B$ inicial es $\vec{0}$ y $B^{-1} = I$, por lo que el resultado es $c$.

En general,

| V.Básica | Ecuación | $\vec{x^*}$ | LD |
|:-:|:-:|:-:|:-:|
| z | F.Obj. | $c^*$ | 0 |
| $\vec{x_B}$ | Rest. | $A^*$ | $b$ |

Ahora "*" es nuestro estandar.

#### Iteración

Se elige el coeficiente más negativo y este entra a la base, entonces cambiemos la columna para la entrada de variable original a la base.

| V.Básica | Ecuación | $\vec{x}$ | LD |
|:-:|:-:|:-:|:-:|
| z | F.Obj. | $c_B B^{-1}A-c$ | $c_B B^{-1}b$ |
| $\vec{x_B}$ | Rest. | $B^{-1} A$ | $B^{-1}b$ |

Entonces, el problema es conseguir $B^{-1}$, aplicamos el método al mismo ejemplo.

### Ejercicio 17

Su empresa vende diariamente dos tipos de computadores, el computador a lo vende a 900 dolares y el b que es la versión barata a 500 dolares. En el armado del computador a utiliza 4 horas y en el computador b usa una hora, tiene un total de 10 horas al día para armar computadores. Usted gasta en componentes 400 dolares para el producto A y 200 dolares para el producto b y tiene para invertir diariamente 1200 dolares. Adicionalmente tiene un mostrador, donde puede poner a lo más 4 computadores, no tiene bodega por lo que es el limite de inventario. Suponga que todo se vende diariamente, formule y solucione el modelo de programación lineal que defina la producción diaria.

#### Algoritmo Simplex Matricial

El modelo del tableau (para max) del problema en este caso.

| Tableau | $x_1$ | $x_2$ | $x_3$ | $x_4$ | $x_5$ | R |
|---|---:|---:|---:|---:|---:|---:|
| CR | 500 | 300 | 0 |0 |0 | 0|
| $x_3$ | 4 | 2 | 1 |0 |0 | 12|
| $x_4$ | 4 | 1 | 0 |1 |0 | 10|
| $x_5$ | 1 | 1 | 0 |0 |1 | 4|

Entonces, la base inicial es $x_3$, $x_4$ y $x_5$, por lo que $B = I$, entonces $B^{-1} = I$.

#### Paso 1 algoritmo:

- Optimalidad: Calcular $c_B B^{-1}A-c$

- $c_B = [0 \quad 0 \quad 0]$
- $B^{-1} = I_3 $
- $A = \begin{bmatrix}
4 & 2 \\
4 & 1 \\
1 & 1
\end{bmatrix}$
- $c = [500 \quad 300]$

Entonces el resultado es:

$c_B B^{-1}A-c = [-500 \quad -300]$, no es óptimo, entra $x_1$ a la base.

$x_4$ sale de la base, del mismo modo que el algoritmo tableau.

#### Paso 2 algoritmo

Actualizamos los calculos. $Base = [x_3, x_1, x_5], B_A = [x_4, x_2]$

- $c_B = [0 \quad 500 \quad 0]$
- $B = \begin{bmatrix}
1 & 4 & 0 \\
0 & 4 & 0 \\
0 & 1 & 1
\end{bmatrix}$
- $A = \begin{bmatrix}
0 & 2 \\
1 & 1 \\
0 & 1
\end{bmatrix}$
- $c = [0 \quad 300]$


In [33]:
import numpy as np
from numpy.linalg import inv

B = np.array([[1, 4, 0], [0, 4, 0], [0, 1, 1]])
Binv = inv(B)
Binv

array([[ 1.  , -1.  ,  0.  ],
       [ 0.  ,  0.25,  0.  ],
       [ 0.  , -0.25,  1.  ]])

In [34]:
A = np.array([[0, 2], [1, 1], [0, 1]])
cb = np.array([[0, 500, 0]])
c = np.array([0, 300])
b = np.array([12, 10, 4])

cb@Binv@A-c

array([[ 125., -175.]])

Como -175 es negativo, no tenemos la condición de optimalidad,  por lo que $x_2$ entra a la base. Veamos quien sale, debemos calcular $B^{-1} A$ y $B^{-1} b$.

In [35]:
Binv@A

array([[-1.  ,  1.  ],
       [ 0.25,  0.25],
       [-0.25,  0.75]])

In [36]:
Binv@b

array([2. , 2.5, 1.5])

In [37]:
cb@Binv@b

array([1250.])

Se saca el minimo entre 2/1, 2.5/0.25, 1.5/0.75. Entonces es 2, 10, 2. Se elige cualquiera, por lo que $x_2$ entra a la base $x_5$ sale de ella.

#### Paso 3 algoritmo

Actualizamos los calculos. $Base = [x_3, x_1, x_2], B_A = [x_4, x_5]$

- $c_B = [0 \quad 500 \quad 300]$
- $B = \begin{bmatrix}
1 & 4 & 2 \\
0 & 4 & 1 \\
0 & 1 & 1
\end{bmatrix}$
- $A = \begin{bmatrix}
0 & 0 \\
1 & 0 \\
0 & 1
\end{bmatrix}$
- $c = [0 \quad 0]$


In [61]:
B = np.array([[1, 4, 2], [0, 4, 1], [0, 1, 1]])
Binv = inv(B)
Binv

array([[ 1.        , -0.66666667, -1.33333333],
       [ 0.        ,  0.33333333, -0.33333333],
       [ 0.        , -0.33333333,  1.33333333]])

In [67]:
A = np.array([[0, 0], [1, 0], [0, 1]])
cb = np.array([[0, 500, 300]])
c = np.array([0, 0])
b = np.array([12, 10, 4])

cb@Binv@A-c

array([[ 66.66666667, 233.33333333]])

Como no hay negativos, cumple optimalidad.

In [68]:
Binv@A

array([[-0.66666667, -1.33333333],
       [ 0.33333333, -0.33333333],
       [-0.33333333,  1.33333333]])

In [69]:
Binv@b

array([8.8817842e-16, 2.0000000e+00, 2.0000000e+00])

In [70]:
cb@Binv@b

array([1600.])

Entonces, la solución es (2,2) y el optimo es 1600.